src/compas/datastructures/mesh/core/mesh.py

```diff
@@ -1371,25 +1375,56 @@ def vertices_on_boundaries(self):
         if not vertices_all:
             return vertices_all
 
-        key = sorted([(key, self.vertex_coordinates(key)) for key in vertices_all], key=lambda x: (x[1][1], x[1][0]))[0][0]
+        special = []
+        for key in vertices_all:
+            count = 0
+            for nbr in self.vertex_neighbors(key):
+                face = self.halfedge_face(key, nbr)
+                if face is None:
+                    count += 1
+            if count > 1:
+                special.append(key)
 
         boundaries = []
-        while vertices_all:
-            vertices = []
-            start = key
-            while 1:
-                for nbr, fkey in iter(self.halfedge[key].items()):
-                    if fkey is None:
-                        vertices.append(nbr)
-                        key = nbr
+
+        if special:
+            for start in special:
+                for key in self.vertex_neighbors(start):
+                    face = self.halfedge_face(start, key)
+                    if face is None:
+                        vertices = [start, key]
+                        while True:
+                            for nbr in self.vertex_neighbors(key):
+                                face = self.halfedge_face(key, nbr)
+                                if face is None:
+                                    if nbr != start:
+                                        vertices.append(nbr)
+                                    key = nbr
+                                    break
+                            if key == start:
```

Comment:
Or maybe consider the more concrete case of a mesh consisting of three triangles, with the middle triangle sharing one vertex with the triangle to the left and one vertex with the triangle to the right, giving two special vertices in the mesh, and all edges are boundary edges.  I think the loops in this case might result in "random" walks around the boundary.

Reply:
yes there are still (many) uncovered cases. perhaps it is better to limit the function to manifold meshes? we could add a separate algorithm that still tries to find boundaries but with the disclaimer that it is to be used "at your own discretion"...

Reply:
Would it makes sense to turn the logic around for these boundary methods?  First one finds all of the faces on the boundary by taking a reference count for each unoriented edge on each face.  The faces attached to edges that have reference count 1 are the boundary faces.  This avoids the non-manifold issues because there are no special faces, and special edges (those that are coincident to three or more faces) are thrown out in the process.  This also finds the collection of boundary edges, from which we want the connected components.  For this, we can start playing dominoes with the edges, by which I mean, we can initialize a boundary component list with a randomly chosen boundary edge and choose a vertex from that edge, perform a breadth-first search, looking through the remaining edges to find ones coincident with that vertex, append those to our current boundary component (and remove them from the "remaining edges"), and then move on to the next edge in the component until we reach the end, and then start again with a new randomly chosen boundary edge.  The order of the edges (and hence vertices) within a component for those without special vertices.  And for those with... well, I don't think order is well-defined.

Reply:
that is actually a very good idea. will try to implement something along those lines...

Reply:
so this didn't actually work but i extended the algorithm based on the separate processing of non-manifold vertices. it now always seems to work (see the tests) ...